backend/src/service/nextcloud.rs

```diff
@@ -0,0 +1,284 @@
+    pub meta: Meta,
+    pub data: CircleData,
+}
+
+#[derive(Debug, Deserialize)]
+pub struct Meta {
+    pub status: String,
+    pub statuscode: i32,
+    pub message: String,
+}
+
+#[derive(Debug, Deserialize)]
+pub struct CircleData {
+    pub id: String,
+    pub name: String,
+}
+
+#[derive(Debug, Deserialize)]
+#[serde(rename_all = "snake_case")]
+pub struct FindCircleData {
+    pub element: FindCircleDataElement,
+}
+
+#[derive(Debug, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct FindCircleDataElement {
+    pub id: String,
+    pub user_id: String,
+    pub user_type: i32,
+    pub display_name: String,
+    pub instance: String,
+}
+
+fn extract_id_from_doc(doc: &str) -> String {
+    let res: CreateCircleResponse = quick_xml::de::from_str(doc).unwrap();
+    log::debug!("{:?}", res);
+    res.data.id
+}
+
+fn extract_id_from_circle_search_response(doc: &str) -> String {
+    let res: FindCircleResponse = quick_xml::de::from_str(doc).unwrap();
+    log::debug!("{:?}", res);
+    res.data.element.id.clone()
+}
+
+pub async fn setup_nextcloud_shares(
+    user_id: Uuid,
+    map_name: &str,
+    token: &str,
+    _should_share_publicly: bool,
+) -> Result<(), ServiceError> {
+    // create Nextcloud circle with the map name
+    let circle_id = create_nextcloud_circle(map_name, token).await?;
+    // create a directory with the map name
+    create_directory(user_id, map_name, token).await?;
+    // create a directory for the base layer within
+    create_directory(user_id, &format!("{}/Base/", map_name), token).await?;
+    share_directory(map_name, token, circle_id).await?;
+
+    // FIXME: nextcloud returns 500 error
+    // if should_share_publicly {
+    //     let public_circle_id = find_circle("PermaplanT", token).await?;
+    //     share_directory(map_name, token, public_circle_id).await?;
+    // }
+
+    Ok(())
+}
+
+#[derive(Debug, Deserialize)]
+#[serde(rename_all = "snake_case")]
+pub struct FindCircleResponse {
+    pub meta: Meta,
+    pub data: FindCircleData,
+}
+
+#[derive(Serialize, Debug)]
+#[serde(rename_all = "camelCase")]
+struct NextcloudSharedOptionsDto {
+    path: String,
+    share_type: i16,
+    permissions: i16,
+    share_with: String,
+}
+
+// Share a Nextcloud directory with a Nextcloud circle
+async fn share_directory(
+    dir_name: &str,
+    token: &str,
+    circle_id: String,
+) -> Result<(), ServiceError> {
+    // Share the map directory with the circle
+    let share_options = NextcloudSharedOptionsDto {
+        path: format!("/PermaplanT/{}", dir_name),
+        share_type: 7,
+        permissions: 31,
+        share_with: circle_id.to_owned(),
+    };
+
+    log::debug!("Sharing directory");
+    log::debug!("{:?}", share_options);
+
+    // For public maps share the map directory with PermaplanT circle
+    let url = BASE_URL.to_owned() + "ocs/v2.php/apps/files_sharing/api/v1/shares";
+    let client = reqwest::Client::new();
+    let res = client
```

Comment:
Instead of recreating the http client each time we want to make a request, we should use a global http client across the whole app. 
In an environment under load, this could otherwise cause socket exhaustion. https://www.aspnetmonsters.com/2016/08/2016-08-27-httpclientwrong/
Even though we are not at that scale yet, it is probably better to already think about it.

Reply:
Great suggestion, thank you! I added this to backend coding guidelines b597457c7b56fb4d0833d484c7744d14dd807eb9